ambry-clustermap/src/main/java/com.github.ambry.clustermap/HelixBootstrapUpgradeUtil.java

```diff
@@ -645,35 +674,41 @@ private void addUpdateInstances(String dcName, Map<String, Set<String>> partitio
   private void addUpdateResources(String dcName, Map<String, Set<String>> partitionsToInstancesInDc) {
     HelixAdmin dcAdmin = adminForDc.get(dcName);
     List<String> resourcesInCluster = dcAdmin.getResourcesInCluster(clusterName);
+    // maxResource may vary from one dc to another (special partition class allows partitions to exist in one dc only)
+    int maxResource = -1;
     for (String resourceName : resourcesInCluster) {
       boolean resourceModified = false;
       if (!resourceName.matches("\\d+")) {
         // there may be other resources created under the cluster (say, for stats) that are not part of the
         // cluster map. These will be ignored.
         continue;
       }
-      maxResource = Math.max(maxResource, Integer.valueOf(resourceName));
+      maxResource = Math.max(maxResource, Integer.parseInt(resourceName));
       IdealState resourceIs = dcAdmin.getResourceIdealState(clusterName, resourceName);
       for (String partitionName : new HashSet<>(resourceIs.getPartitionSet())) {
         Set<String> instanceSetInHelix = resourceIs.getInstanceSet(partitionName);
         Set<String> instanceSetInStatic = partitionsToInstancesInDc.remove(partitionName);
         if (instanceSetInStatic == null || instanceSetInStatic.isEmpty()) {
           if (forceRemove) {
-            info("*** Partition {} no longer present in the static clustermap, removing from Resource *** ",
-                partitionName);
+            info("[{}] *** Partition {} no longer present in the static clustermap, removing from Resource *** ",
+                dcName.toUpperCase(), partitionName);
             // this is a hacky way of removing a partition from the resource, as there isn't another way today.
             // Helix team is planning to provide an API for this.
             resourceIs.getRecord().getListFields().remove(partitionName);
             resourceModified = true;
           } else {
             info(
-                "*** forceRemove option not provided, resources will not be removed (use --forceRemove to forcefully remove)");
+                "[{}] *** forceRemove option not provided, resources will not be removed (use --forceRemove to forcefully remove)",
+                dcName.toUpperCase());
             expectMoreInHelixDuringValidate = true;
-            partitionsNotForceRemovedByDc.computeIfAbsent(dcName, k -> new HashSet<>()).add(partitionName);
+            partitionsNotForceRemovedByDc.computeIfAbsent(dcName, k -> ConcurrentHashMap.newKeySet())
+                .add(partitionName);
           }
         } else if (!instanceSetInStatic.equals(instanceSetInHelix)) {
-          info("Different instance sets for partition {} under resource {}. Updating Helix using static.",
-              partitionName, resourceName);
+          info(
+              "[{}] Different instance sets for partition {} under resource {}. Updating Helix using static. Previous instance set: [{}], new instance set: [{}]",
```

Comment:
Too long, better to split it.

Reply:
Sorry, I don't quite understand, could you offer some suggestions here? Thanks in advance.

Reply:
I mean this line or string is too long. It is better to split into two lines.